packages/next/src/server/app-render/validate-url.ts

```diff
@@ -1,7 +1,7 @@
 const DUMMY_ORIGIN = 'http://n'
 const INVALID_URL_MESSAGE = 'Invalid request URL'
 
-export function validateURL(url: string | undefined): string {
+export function validateURL(url: string | undefined): URL {
```

Comment:
would be better IMO to type this as `IncomingMessage` since it's only used in one place and it's not a safe general purpose URL validator.

We can also get rid of the origin checks because the IncomingMessage url does not contain a host value. I'm curious why this was added in the original implementation. Perhaps there are bugs in Node or this code was being used for Edge and we had a fake IncomingMessage that didn't respect the same constraints that Node does? If that is the case we maybe ought to loosen the types further so we don't operate with a false sense of security that the req is actually an IncomingMessage object.

And given this method currently uses a dummy origin we probably shouldn't return a URL since it's not representative of anything real. We currently only use it for pathname, might as well just have it return the pathname for now

Reply:
I read the tests showing why the canonical url check exists and makes sense so we wouldn't want to get rid of that afterall. but I think the typing, naming, and return value point still stands